### PR TITLE
[SPARK-42840][SQL] Change `_LEGACY_ERROR_TEMP_2004` error to internal error 

### PR DESCRIPTION
**Pull request by Leibnizhu**
### What changes were proposed in this pull request?

In the PR, I propose to change the `_LEGACY_ERROR_TEMP_2004`  error to an internal error. Also this PR improves the error message.

### Why are the changes needed?

`_LEGACY_ERROR_TEMP_2004` cannot be triggered from user code (for instance, some SQL query), more detail:

1. `_LEGACY_ERROR_TEMP_2004` error is thrown in `Literal.default` for `CharType`, `VarcharType` and other user-implemented DataType.
2. `Literal.default` is called in these cases below:
    1. in `org.apache.spark.sql.catalyst.expressions.aggregate.Average`, for getting initial value; but in this case, DataType would only be  DecimalType / YearMonthIntervalType / DayTimeIntervalType / DoubleType; these types are supported by `Literal.default`.
    2. in `org.apache.spark.sql.catalyst.expressions.aggregate.Sum`, is same as `Average`
    3. in `org.apache.spark.sql.catalyst.planning.ExtractEquiJoinKeys` and `org.apache.spark.sql.catalyst.plans.logical.AsOfJoin`; In real scene, they are use for DateFrame api, but `CharType` and `VarcharType` cannot be as a part of DataFrame's schema( they would be converted to StringType); and user-defined DataType do not have a matched Encoder in `org.apache.spark.sql.catalyst.encoders.RowEncoder#encoderForDataType`

so, it should be an internal error, but not `SparkRuntimeException`.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

By running the modified test suites:

```bash
 build/sbt "sql/testOnly org.apache.spark.sql.catalyst.expressions.LiteralExpressionSuite"
```

### CODE REVIEW

**Comment by Leibnizhu** [comment]
@MaxGekk Plz review this PR, thx

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/LiteralExpressionSuite.scala [hunk 1]
```diff
@@ -88,6 +88,17 @@ class LiteralExpressionSuite extends SparkFunSuite with ExpressionEvalHelper {
     checkEvaluation(Literal.default(StructType(StructField("a", StringType) :: Nil)), Row(""))
     // ExamplePointUDT.sqlType is ArrayType(DoubleType, false).
     checkEvaluation(Literal.default(new ExamplePointUDT), Array())
+
+    // DateType without default value`
+    List(CharType(1), VarcharType(1)).foreach(errType => {
+      checkError(
+        exception = intercept[SparkRuntimeException] {
+          Literal.default(errType)
```
[the remark this answers is not in the document]
Comment (Leibnizhu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/LiteralExpressionSuite.scala:97]: fixed

**Review by MaxGekk — CHANGES_REQUESTED** [review]

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
@Leibnizhu Could you fix PR's description and title according to your actual changes, please.

**Comment by Leibnizhu** [comment]
> @Leibnizhu Could you fix PR's description and title according to your actual changes, please.

done

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @Leibnizhu.

**Comment by MaxGekk** [comment]
@Leibnizhu Congratulations with your first contribution to Apache Spark!

**Comment by Leibnizhu** [comment]
> @Leibnizhu Congratulations with your first contribution to Apache Spark!

thx 😁😁😁

**Pull request closed without merging**